### PR TITLE
[PyTorch] Add IValue::IValue(std::vector<T>&&) ctors

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117769

There are two IValue constructors that take `const std::vector<T>&`. Add moving variants to allow callers to save on reference counting.

Differential Revision: [D52879065](https://our.internmc.facebook.com/intern/diff/D52879065/)

### CODE REVIEW

**Review by suo — APPROVED** [review]
okie doke

File: aten/src/ATen/core/ivalue_inl.h [hunk 1]
```diff
@@ -2193,6 +2210,22 @@ inline IValue::IValue(const std::vector<T>& v) : IValue(c10::List<T>()) {
...
   }
 }
+
+template <class T, IValue::enable_if_list_is_ivalue_constructible<T>>
+inline IValue::IValue(std::vector<T>&& v) : IValue(c10::List<T>()) {
+  auto list = to<c10::List<T>>();
+  list.reserve(v.size());
+  if constexpr (std::is_same_v<T, bool>) {
+    for (auto e : v) {
+      list.push_back(e);
```
Comment (Skylion007) [inline aten/src/ATen/core/ivalue_inl.h:2220]:
```suggestion
      list.emplace_back(std::move(e));
```

File: aten/src/ATen/core/ivalue_inl.h [hunk 2]
```diff
@@ -2193,6 +2210,22 @@ inline IValue::IValue(const std::vector<T>& v) : IValue(c10::List<T>()) {
...
+    }
+  } else {
+    for (auto& e : v) {
+      list.push_back(std::move(e));
```
Comment (Skylion007) [inline aten/src/ATen/core/ivalue_inl.h:2224]:
```suggestion
      list.emplace_back(std::move(e));
```

**Review by Skylion007 — APPROVED** [review]
Thanks, this has been on my roadmap for a while.

**Pull request closed without merging**